In [59]:
from __future__ import unicode_literals

import pandas as pd
import numpy as np
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

# 1. Preparing data - 28 weeks of life of the banana plant
---

In [60]:
precipitations_df = pd.read_csv('../../../../data/raw/FincaPorvenir/Metereologico/28-weeks_January-12_August-03_2018/' \
                                 'Precipitaciones_January-12_August-03_2018.csv', )

In [61]:
print(precipitations_df.shape)
precipitations_df.head()

(9750, 2)


Fecha:  Precipitación (ml)
0  2018-01-12 00:17:28                0.00
1  2018-01-12 00:47:28                0.00
2  2018-01-12 01:17:27                0.00
3  2018-01-12 01:47:27                0.00
4  2018-01-12 02:17:27                3.19

## 1.1. We evaluate if this dataset has null type `NaN`
---

In [62]:
print(precipitations_df.isnull().any())
precipitations_df.isnull().values.any()

Fecha:                False
Precipitación (ml)    False
dtype: bool


False

Don't have null values

## 1.2. Selecting  relevant index columns features
---

Since the dataset has a column called **`Fecha:`**, which is not a numerical value,

it will be removed so that it does not interfere **with our subsequent scaling**, 

so we are only going to reference the values or samples of the column

** `Precipitación (ml)` ** and assigning them to the matrix 

`precipitations_array` created such as follow:

In [45]:
precipitations_array = precipitations_df.iloc[:, 1].values

In [63]:
# Rehape the luminosity_luxes array 
precipitations_array = precipitations_array.reshape(-1, 1)
precipitations_array

array([[0.],
       [0.],
       [0.],
       ...,
       [0.],
       [0.],
       [0.]])

## 1.3 Generating descriptive data to dataset
---

In [64]:
col = ['Precipitación (ml)']
precipitations_df = pd.DataFrame(precipitations_array, columns=col)
precipitations_df_describe = precipitations_df.describe()

In [65]:
precipitations_df_describe


Precipitación (ml)
count         9750.000000
mean             0.123362
std              1.224794
min              0.000000
25%              0.000000
50%              0.000000
75%              0.000000
max             44.000000

In [66]:
# Export this descriptive data to comma separated values and java script object notation
precipitations_df_describe.to_csv('../../../../data/interim/Precipitations/28-weeks_January-12_August-03_2018/' +'\n' 
                              'Precipitations_Describe_January-12_August-03.cvs', sep=',', header=True, index=True)
precipitations_df_describe.to_json('../../../../data/interim/Precipitations/28-weeks_January-12_August-03_2018/' +'\n'
                                     'Precipitations_Describe_January-12_August-03.json')

# 2. Creating Precipitations Training and Testing datasets

We have a **`precipitations_df`** dataset with 9750 samples rows.


In [69]:
print(precipitations_array.shape)
# precipitations_df.head()

(9750, 1)


We'll divide it into two differents datasets:

- Training dataset
- Testing dataset

This is executed through [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html  "sklearn.model_selection.train_test_split") function of this way:

`train_test_split` receive as a data parameter a numpy array, we have to turn the 

**`precipitations_df`** dataframe to numpy array such as follow:

In [71]:
# numpy_precipítations = precipitations_df.reset_index().values
# numpy_precipítations = precipitations_df.values
precipitations_array.shape

(9750, 1)

In [72]:
# My numpy_precipítations variable now is a numpy array
# print(numpy_precipítations.shape)
# print(type(numpy_precipítations))
# numpy_precipítations
precipitations_array

array([[0.],
       [0.],
       [0.],
       ...,
       [0.],
       [0.],
       [0.]])

We compose the following datasets from **`precipitations_array`** array :

- `precipitations_train`, which is the training matrix
- `precipitations_test`, the testing matrix

We use the [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html  "sklearn.model_selection.train_test_split") function to create the training and testing dataset. Their parameters are:

![alt text](https://cldup.com/hukPWvvLxt-3000x3000.png "klearn.model_selection.train_test_split")


- The first parameter **should be an array**, then we pass the **`precipitations_array`** which contain all column features of luminosity.

Of this way, we pass all data (9750 samples rows) to from them `precipitations_train`

and `precipitations_test` will be created

- The `test_size=0.5` parameter means a 50% division; which means 
that half of the data goes to the test dataset and the other half 
goes to the training dataset.

A good choice for the size of tests **is usually 0.2 ie 20% or 0.25 or even 30%.** 

In some rare cases we will have 40% but almost never 0.5 or 50%

We choose 20% which means that we will have 20% of 9752 samples or observations for the test data set, 

in this case **9750 * 0.2 = 1950 samples or records for the test data set**


- The `train_size` parameter is the training dataset size.  **`test_size + train_size = 1 or 100%`**, 
then isn't necessary include it, because wheter we include to `test_size = 0.2`, then the remaining 
data will be to `train_size` this means **0.8 or 80%**

This means that **9750 * 0.8 = 78006  training dataset samples rows. **


- random_state is a seed or data source for generating random values for the data sets. 

If this parameter is not passed, the data will be generated in a random way, but in the way as by default numpy works them.

### 2.1  Creating Training and testing precipitations datasets
---

- `precipitations_train`, which is the training matrix
- `precipitations_test`, the testing matrix

In [73]:
precipitations_train, precipitations_test = train_test_split(precipitations_array, test_size = 0.2)

In [74]:
# We have 7800 rows to luminosity_luxes_train
print(type(precipitations_train))
print("The dimensionality of wind_direction training dataset is: " +'\n' , precipitations_train.shape)
print('\n')

# And we have 1950 rows to luminosity_luxes_test
print(type(precipitations_test))
print("The dimensionality of wind_direction testing dataset is: " +'\n' , precipitations_test.shape)

<class 'numpy.ndarray'>
The dimensionality of wind_direction training dataset is: 
 (7800, 1)


<class 'numpy.ndarray'>
The dimensionality of wind_direction testing dataset is: 
 (1950, 1)


In this way, the model, as you progressively learn the correlations in the training set, the better the prediction of the results in the test set.

But if the model learns too much from memory the correlations of the training sets, 

that is to say, when one learns from memory and does not understand things, then it 

will have problems to predict what is happening on the set of tests, because it is 

learned for difficult correlations, if the logic is not well understood and you can 

not make good predictions. This is called overfitting or overfitting


The really important thing is to understand that we need to have two different datasets

- Training set with which the ML model learns
- Test set, on which we test whether the ML model correctly learned the correlations

---
##  3. Feature Scaling Precipitations training and testing dataset  
---

[This article post](http://benalexkeen.com/feature-scaling-with-scikit-learn/  "Feature Scaling with scikit-learn") it's a great reference to explore the features scaling methods
on scikit learn

- `StandardScaler` assume that data is normally distributed at the level of each characteristic or variable. If the data is not normally distributed, it is not the best alternative to use for scaling.  

- `Min-Max Scaler` it is probably the most famous scaling algorithm and what it does is resize the range to leave it in a dimension of 0 to 1 or -1 and 1 (in case there are negative values in the original dataset of input data)

This scale of maximums and minimums works best for cases where standard scaling may not work properly. If the distribution is not Gaussian or the standard deviation is very small the escalation of maximums and minimums is the best idea.
However, it is sensitive to outliers or outliers, so if there are outliers in the data it is better to consider robust scaling.

- `Robust Scaler` it is similar to the previous one of maximums and minimums, only that it uses interquartile ranges instead of maximums and minimums, which makes it robust for the outliers

- `Normalizer` which scales each value, dividing each value by its magnitude in n dimensional spaces for n number of characteristics.


---



We will use the scaling of maximums and minimums to scale the water table data, because the standard deviation is very small, 

it does not have atypical values and it does not follow a normal distribution (you have to check this)

We apply the maximum and minimum scaling. We provide a rank or base scale that **will be between 0 and 1** using an object,

[MinMaxScaler](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html "sklearn.preprocessing.MinMaxScaler") which transforms each characteristic, (in this 

case it will be the columns feature of `Precipitación (ml)`) individually according to a given range.

Product of its applicability, generates these attributes in the dataset, already transformed: 

![alt text](https://cldup.com/lTIv4HXgTk-3000x3000.png "sklearn.preprocessing.MinMaxScaler")

### 3.1 We apply maximium and minimum feature scaling to Precipitations training dataset

In [76]:
# We provide a base scale range
scaler = MinMaxScaler(feature_range=(0, 1))

print("Remember our precipitations training data " + '\n', precipitations_train)

Remember our precipitations training data 
 [[0.]
 [0.]
 [0.]
 ...
 [0.]
 [0.]
 [0.]]


With the [fit](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler.fit  "MinMaxScaler.fit")
we compute the maximum and minimum value of  `precipitations_train` dataset to be used in the subsequent scaling 

We assing these values to `minmax_scale_training` variable.

In [77]:
minmax_scale_training = scaler.fit(precipitations_train.astype(float))
# print(minmax_scale_training.data_max_)
# http://terrapinssky.blogspot.com/2017/10/pythonresolved-dataconversionwarning.html

Then, we apply the [transform](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler.transform "MinMaxScaler.transform") method to transform these data to maximum and mínimum scale value. 

Here, with this process, the `precipitations_train` data are scaled between **0 to 1**  selected range 


In [78]:
# transform precipitations_train data to maximum and mínimum scale value. 
precipitations_minmax_training = minmax_scale_training.transform(precipitations_train)

In [79]:
print ("And, these are our scaled data: " + '\n')
precipitations_minmax_training

And, these are our scaled data: 



array([[0.],
       [0.],
       [0.],
       ...,
       [0.],
       [0.],
       [0.]])

In [90]:
print('Precipitations Training dataset. Minimum value after MaxMinScaler:\nLuminosity={:.1f}'
      .format(precipitations_minmax_training[:,0].min()))

print('Precipitations Training dataset. Maximum value after MaxMinScaler:\nLuminosity={:.1f}'
      .format(precipitations_minmax_training[:,0].max()))

Precipitations Training dataset. Minimum value after MaxMinScaler:
Luminosity=0.0
Precipitations Training dataset. Maximum value after MaxMinScaler:
Luminosity=1.0


Then, our  MinMaxScaler normalized training dataset is `precipitations_minmax_training` numpy array

- We export this array to comma separated values 


In [57]:
precipitations_train_df = pd.DataFrame(precipitations_minmax_training, columns=col)

In [58]:
precipitations_train_df

Precipitación (ml)
0               0.000000
1               0.000000
2               0.000000
3               0.000000
4               0.000000
5               0.000000
6               0.000000
7               0.000000
8               0.000000
9               0.000000
10              0.000000
11              0.000000
12              0.000000
13              0.000000
14              0.000000
15              0.000000
16              0.000000
17              0.000000
18              0.000000
19              0.000000
20              0.000000
21              0.054924
22              0.000000
23              0.000000
24              0.000000
25              0.000000
26              0.000000
27              0.000000
28              0.000000
29              0.000000
...                  ...
7770            0.000000
7771            0.000000
7772            0.000000
7773            0.000000
7774            0.000000
7775            0.000000
7776            0.000000
7777            0.041193
7778            0.000000
7779            0.000000
7780            0.000000
7781            0.000000
7782            0.000000
7783            0.000000
7784            0.000000
7785            0.000000
7786            0.000000
7787            0.000000
7788            0.000000
7789            0.000000
7790            0.000000
7791            0.000000
7792            0.000000
7793            0.000000
7794            0.000000
7795            0.000000
7796            0.000000
7797            0.000000
7798            0.000000
7799            0.000000

[7800 rows x 1 columns]

In this way we have the dataset `precipitations_train_df` standardized and training, and export it to a .csv file

In [80]:
precipitations_train_df.to_csv('../../../../data/processed/Precipitations/28-weeks_January-12_August-03_2018/' +'\n' 
                                 'Precipitations_Normalized_TRAINING_January-12_August-03.csv', sep=',', header=True, index=False)

### 3.2 We apply maximium and minimum feature scaling to Precipitations testing dataset

In [81]:
# We provide a base scale range
scaler = MinMaxScaler(feature_range=(0, 1))

print("Remember our precipitations testing data " + '\n', precipitations_test)

Remember our luminosity testing data 
 [[0.]
 [0.]
 [0.]
 ...
 [0.]
 [0.]
 [0.]]


In [82]:
minmax_scale_test = scaler.fit(precipitations_test.astype(float))
# transform luminosity_luxes_test data to maximum and mínimum scale value. 
precipitations_minmax_test = minmax_scale_test.transform(precipitations_test)

In [83]:
print ("And, these are our testing scaled data: " + '\n')
precipitations_minmax_test

And, these are our testing scaled data: 



array([[0.],
       [0.],
       [0.],
       ...,
       [0.],
       [0.],
       [0.]])

In [84]:
print('Luminosity Testing dataset. Minimum value after MaxMinScaler:\nLuminosity={:.1f}'
      .format(precipitations_minmax_test[:,0].min()))

print('Luminosity Testing dataset. Maximum value after MaxMinScaler:\nLuminosity={:.1f}'
      .format(precipitations_minmax_test[:,0].max()))

Luminosity Testing dataset. Minimum value after MaxMinScaler:
Luminosity=0.0
Luminosity Testing dataset. Maximum value after MaxMinScaler:
Luminosity=1.0


Then, our  MinMaxScaler normalized training dataset is `precipitations_minmax_test` numpy array

- We export this array to comma separated values

In [86]:
precipitations_test_df = pd.DataFrame(precipitations_minmax_test, columns=col)

In [87]:
print(precipitations_test_df.shape)
precipitations_test_df.head()

(1950, 1)


Precipitación (ml)
0                 0.0
1                 0.0
2                 0.0
3                 0.0
4                 0.0

In [ ]:
precipitations_test_df.to_csv('../../../../data/processed/Luminosity/28-weeks_January-12_August-03_2018/' +'\n' 
                                 'Luminosity_Normalized_TESTING_January-12_August-03.csv', sep=',', header=True, index=False)